### PR TITLE
[NTOS:MM] Add locking around MmLocateMemoryAreaByAddress

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Improve locking for MmLocateMemoryAreaByAddress._

JIRA issue: [CORE-18519](https://jira.reactos.org/browse/CORE-18519) (fixed already after the 1st commit in here)

## Proposed changes
```
Try to improve compiling ReactOS on ReactOS after receiving the below on MSVC build:
nt!MiCheckForConflictingNode+0x8b:
804f1ccb 3b5110          cmp     edx,dword ptr [ecx+10h]
kd> g
Access violation - code c0000005 (!!! second chance !!!)
nt!MiCheckForConflictingNode+0x8b:
804f1ccb 3b5110          cmp     edx,dword ptr [ecx+10h]
kd> k
 # ChildEBP RetAddr  
00 f6b60ca8 80502a36 nt!MiCheckForConflictingNode+0x8b [d:\reacto\reactos\ntoskrnl\mm\arm3\vadnode.c @ 96]
01 f6b60cec 80502c8a nt!MmLocateMemoryAreaByAddress+0x66 [d:\reacto\reactos\ntoskrnl\mm\marea.c @ 75]
02 f6b60d10 80565c8a nt!MmAccessFault+0x8a [d:\reacto\reactos\ntoskrnl\mm\mmfault.c @ 242]
03 f6b60d5c 804036fe nt!KiTrap0EHandler+0x13a [d:\reacto\reactos\ntoskrnl\ke\i386\traphdlr.c @ 1369]
04 f6b60d5c 7c93e90f nt!KiTrap0E+0x99
05 0245f554 7c93dfa4 ntdll!RtlpFindAndCommitPages+0x5bf [d:\reacto\reactos\sdk\lib\rtl\heap.c @ 814]
06 0245f5c0 7c93cc8b ntdll!RtlpExtendHeap+0x84 [d:\reacto\reactos\sdk\lib\rtl\heap.c @ 1279]
07 0245f650 7c9399eb ntdll!RtlpAllocateNonDedicated+0x8b [d:\reacto\reactos\sdk\lib\rtl\heap.c @ 1971]
08 0245f744 7c5c2364 ntdll!RtlAllocateHeap+0x2eb [d:\reacto\reactos\sdk\lib\rtl\heap.c @ 2132]
09 0245f764 7c5c22a8 msvcrt!msvcrt_heap_alloc+0x84 [d:\reacto\reactos\sdk\lib\crt\wine\heap.c @ 93]
*** WARNING: Unable to verify timestamp for cc1.exe
*** ERROR: Module load completed but symbols could not be loaded for cc1.exe
```
Note: The associated Jira ticket has been updated.

### CODE REVIEW

**Comment by JoachimHenze** [comment]
I retested this patch against CORE-18516 but I had no success to complete the build-process, see https://jira.reactos.org/browse/CORE-18516?focusedId=140170&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-140170 for details and logs. It stopped while logging bullshit at roughly where it always stopped with logging bullshit since we got the new Memory Manager during 0.4.15-dev'ing.

I could not spot anything yet, that would be improved by PR5658. My current opinion is to not merge it (yet).

File: ntoskrnl/mm/mmfault.c
Comment (tkreuzer) [inline ntoskrnl/mm/mmfault.c:243]: Can we have an ASSRT in MmLocateMemoryAreaByAddress?
Something like
`ASSERT(PsGetCurrentProcess()->AddressCreationLock.Owner != KeGetCurrentThread());`

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: @tkreuzer, Sorry, I do not understand your question. Are you asking me to add this statement into the code at MmLocateMemoryAreaByAddress? Or was this question for another developer? I need help here. Thanks.

**Comment by Doug-Lyons** [comment]
> I retested this patch against CORE-18516 but I had no success to complete the build-process, see https://jira.reactos.org/browse/CORE-18516?focusedId=140170&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-140170 for details and logs. It stopped while logging bullshit at roughly where it always stopped with logging bullshit since we got the new Memory Manager during 0.4.15-dev'ing.
> 
> I could not spot anything yet, that would be improved by PR5658. My current opinion is to not merge it (yet).

Sorry, @JoachimHenze I thought that both Ros On Ros tickets would be improved by this patch, but apparently only the one for compiling the current master on current master was improved. I changed the linked Jira issue accordingly. Thanks for testing.

File: ntoskrnl/mm/mmfault.c
Comment (tkreuzer) [reply ntoskrnl/mm/mmfault.c:243]: The first. I formulated it as a question, because I don't know if that will actually work. But please try that.

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: I added what you see in the screen capture below. On boot there is a constant stream of this assert being hit. The debug log is attached for your information as well. I was not able to get past the first black screen. Let me know if you need anything else. Thanks.
marea patch-01
[PR-5658-test.txt](https://github.com/reactos/reactos/files/12682433/PR-5658-test.txt)

Comment (HBelusca) [reply ntoskrnl/mm/mmfault.c:243]: @Doug-Lyons When you get this assertion hit, could you display in the windbg the value of Process, Process->AddressCreationLock (and the Owner field as well) ?
```
kd> ??Process
...
kd> ??Process->AddressCreationLock
...
```
etc.

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: @HBelusca, See Below.
[PR-5658-test-02.txt](https://github.com/reactos/reactos/files/12683691/PR-5658-test-02.txt)

Comment (HBelusca) [reply ntoskrnl/mm/mmfault.c:243]: OK so we have for example,
```
kd> ?? Process
struct _EPROCESS * 0xb13f9bb0
[...]
   +0x094 UniqueProcessId  : 0x00000004 Void
[...]
   +0x164 ImageFileName    : [16]  "System"
[...]
kd> ?? Process->AddressCreationLock
struct _KGUARDED_MUTEX
   +0x000 Count            : 0n1
   +0x004 Owner            : (null) 
   +0x008 Contention       : 0
   +0x00c Gate             : _KGATE
   +0x01c KernelApcDisable : 0n0
   +0x01e SpecialApcDisable : 0n-1
   +0x01c CombinedApcDisable : 0xffff0000
kd> ?? Process->AddressCreationLock.Owner
struct _KTHREAD * 0x00000000
```

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: @HBelusca, after seeing the the .Owner was NULL, I modified the IF to what is shown in the screen shot below and still had the assert triggered as shown in the debug log below.
marea patch-02
[PR-5658-test-03.txt](https://github.com/reactos/reactos/files/12683906/PR-5658-test-03.txt)

**Comment by JoachimHenze** [comment]
I invested a week into testing this PR, and the result can be found here: https://jira.reactos.org/browse/CORE-18519?focusedId=140285&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-140285
(That was without the recently discussed added Timo-Kreiuzer-assert).
In a few days up to 1week I will decide whether to approve this PR! Please wait so long with the merging!

File: ntoskrnl/mm/mmfault.c
Comment (tkreuzer) [reply ntoskrnl/mm/mmfault.c:243]: For the case of `Process == NULL`, you can add `ASSERT(PsIdleProcess->AddressCreationLock.Owner == KeGetCurrentThread());`

**Review by tkreuzer — APPROVED** [review]
Please also add the 2nd ASSERT for Process == NULL, otherwise LGTM.

File: ntoskrnl/mm/mmfault.c
Comment (tkreuzer) [reply ntoskrnl/mm/mmfault.c:243]: I had a look and found missing locks in `MiProtectVirtualMemory` and `MmAccessFault`. So you need to fix those 2 instance, too.

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: @tkreuzer, Please review these changes and see if they are what you had in mind. Thanks lots.

**Review by JoachimHenze — APPROVED** [review]
I completed the test of **the first commit of PR5658**. It succeeded to fix CORE-18519.
Vanilla master confirmed to still fail for CORE-18519.
See https://jira.reactos.org/browse/CORE-18519?focusedId=140285&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-140285
for details.

Since my test 2 more commits have been added (which I didn't test yet. I will redo the test for https://jira.reactos.org/browse/CORE-18516 now with all 3 commits, because it failed very similarly, but that ticket wasn't fixed by just the first commit in here.

**Comment by JoachimHenze** [comment]
I finished the retest of the https://jira.reactos.org/browse/CORE-18516 .
Unfortunately for the RosBE2.1.6 case the compiler segmentation faults are still not fixed **even with all 3 commits of this PR**, see
https://jira.reactos.org/browse/CORE-18516?focusedId=140297&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-140297

But that doesn't mean the PR would be wrong, or I would want to withdraw my approval.
The first commit in here does definitely fix [https://jira.reactos.org/browse/CORE-18519]https://jira.reactos.org/browse/CORE-18519 .
I cannot say, whether there are any testcases that woukd be improved though by the 2nd and 3rd commit (Timos review). If there are any, I didn't find them yet.

Are we good to commit this PR now?

File: ntoskrnl/mm/mmfault.c
Comment (tkreuzer) [reply ntoskrnl/mm/mmfault.c:243]: Yes looks good. Do you still get any assertion failures?

Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: I have only tested booting and running RAPPS to download Total Commander and this happens without triggering any asserts. I should probably do a full compile of ReactOS on ReactOS and then a set of regression tests to say for sure that there are no asserts. I will try and do these things later today. Thanks for your help and attention to detail with this.

Comment (tkreuzer) [reply ntoskrnl/mm/mmfault.c:243]: Can you also add the
```
    else
        ASSERT(PsIdleProcess->AddressCreationLock.Owner == KeGetCurrentThread());
```

**Comment by tkreuzer** [comment]
Tests
x86 looks ok: https://reactos.org/testman/compare.php?ids=89801,89804
x64 doesn't seem to be happy: https://build.reactos.org/#/builders/12/builds/2259

**Comment by Doug-Lyons** [comment]
> Tests x86 looks ok: https://reactos.org/testman/compare.php?ids=89801,89804 x64 doesn't seem to be happy: https://build.reactos.org/#/builders/12/builds/2259

@tkreuzer, I tried reverting the last commit by just commenting the added lines and re-triggering your KVM_x64 build and this still failed.
So I reverted the next-to-least commit as well and re-triggered your KVM_x64 build and this still failed.
I thought that this would have built correctly, because you had used 'branch' "refs/pull/5658/merge".
I cannot trigger a clean build on this bot because I do not have premissions. I wanted to try this;
'mmfault_c_lock_fix-02.patch JID66303 on top of 0.4.15-dev-6661-gcc6e444' which would have been the minimum patch.
I triggered it on the [Build GCCLin_x86](https://build.reactos.org/#/builders/9) a bit ago to see its results.
Maybe you can check behind me and see what you can do. Thanks for your help.

File: ntoskrnl/mm/mmfault.c
Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: @tkreuzer. I am waiting to make this commit until we can get the KVM_x64 bot to run. Thanks.

**Comment by JoachimHenze** [comment]
image
seems to succeed on current master head (without this PR).
What makes you think that this PR would break that bot?
To me it looks as if the link with the broken KVM x64 tester is unrelated to that PR, because it was just flaky (I have seen it failing now and then even without this PR).

I do see no need for anything like:
image

Those 2 commits now make the PR unmergable, because it is castrated.

**Comment by tkreuzer** [comment]
@Doug-Lyons I found the issue. x64 can currently incur a recursive page fault when resolving kernel section mappings. This would lead to recursively acquiring the lock, causing an assertion failure. I have a tiny hack to fix this, which will make your code work: https://github.com/reactos/reactos/pull/5741

**Comment by Doug-Lyons** [comment]
> @Doug-Lyons I found the issue. x64 can currently incur a recursive page fault when resolving kernel section mappings. This would lead to recursively acquiring the lock, causing an assertion failure. I have a tiny hack to fix this, which will make your code work: #5741

@tkreuzer, That's great! :) I will go ahead and fix this PR by putting back the remarked out code where I was trying to get the KVM_x64 bot to work. Thanks Lots!

File: ntoskrnl/mm/mmfault.c
Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: Sorry, @tkreuzer after looking more closely at your request, I am not sure where to add this "else...". Can you be more specific? I have removed the comment slashes now. Thanks.

**Comment by tkreuzer** [comment]
PR #5741 is merged.

File: ntoskrnl/mm/marea.c
Comment (tkreuzer) [inline ntoskrnl/mm/marea.c:76]:
```suggestion
    if (Process == NULL)
        ASSERT(PsIdleProcess->AddressCreationLock.Owner == KeGetCurrentThread());
    else
        ASSERT(Process->AddressCreationLock.Owner == KeGetCurrentThread());
```

Comment (Doug-Lyons) [reply ntoskrnl/mm/marea.c:76]: Changed. Thanks.

File: ntoskrnl/mm/mmfault.c
Comment (Doug-Lyons) [reply ntoskrnl/mm/mmfault.c:243]: Should be fixed now. Thanks.

**Comment by tkreuzer** [comment]
GCCLin_x86 on Test KVM: https://reactos.org/testman/compare.php?ids=90045,90058,90061,90050
MSVC_x64 on Test KVM_x64: https://reactos.org/testman/compare.php?ids=90030,90036,90048,90052

Some weird results on x86 (regedit test regressed?) Gonna run again.

**Comment by Doug-Lyons** [comment]
> GCCLin_x86 on Test KVM: https://reactos.org/testman/compare.php?ids=90045,90058,90061,90050 MSVC_x64 on Test KVM_x64: https://reactos.org/testman/compare.php?ids=90030,90036,90048,90052
> 
> Some weird results on x86 (regedit test regressed?) Gonna run again.

I noticed a very bad results running the 'kernel32 virtual' test locally. 
I was able to continue it with 'O' for once and get a backtrace and then 'cont' as you can see here:
[kernel32-15-6679-06.txt](https://github.com/reactos/reactos/files/12841359/kernel32-15-6679-06.txt)
But eventually it hangs and when that happens, there is no getting the machine to shutdown without a power off.

If you look at this patch, you can remove four (4) lines what I have commented here and all seems to work.
[mmfault-all-works-01.patch](https://github.com/reactos/reactos/files/12841353/mmfault-all-works-01.patch)

**Comment by SergeGautherie** [comment]
'This branch has conflicts that must be resolved':
probably due to PR 5782 merge...

**Comment by Doug-Lyons** [comment]
Superseded by PR #5782.

**Pull request closed without merging**
